cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -579,7 +581,16 @@ class BackpressureHandler {
     }
   }
 
+  Status ForceShutdown() {
+    // It may be unintuitive to call Resume() here, but this is to avoid a deadlock.
+    // Since acero's executor won't terminate if any one node is paused, we need to
+    // force resume the node before stopping production.
+    backpressure_control_->Resume();
```

Comment:
Perhaps one thing to clarify is whether `ResumeInput` behaves idempotently? I.e., is it OK to always call resume, even though only some inputs hit this `PauseInput` race condition?

My perusal of `source_node.cc` tells me this is OK, but LMK if this is a poor assumption to make.